overlord/snapstate/snapstate_update_test.go

```diff
@@ -6844,6 +6845,139 @@ func (s *snapmgrTestSuite) TestUpdatePrerequisiteWithSameDeviceContext(c *C) {
+
+	restore := snapstate.MockEnforcedValidationSets(func(st *state.State) (*snapasserts.ValidationSets, error) {
+		vs := snapasserts.NewValidationSets()
+		snap1 := map[string]interface{}{
+			"id":       "aaqKhntON3vR7kwEbVPsILm7bUViPDzx",
+			"name":     "some-snap",
+			"presence": "required",
+		}
+		snap2 := map[string]interface{}{
+			"id":       "bgtKhntON3vR7kwEbVPsILm7bUViPDzx",
+			"name":     "some-other-snap",
+			"presence": "required",
+		}
+		vsa1 := s.mockValidationSetAssert(c, "bar", "2", snap1, snap2)
+		vs.Add(vsa1.(*asserts.ValidationSet))
+		return vs, nil
+	})
+	defer restore()
+
+	s.state.Lock()
+	defer s.state.Unlock()
+
+	tr := assertstate.ValidationSetTracking{
+		AccountID: "foo",
+		Name:      "bar",
+		Mode:      assertstate.Enforce,
+		Current:   2,
+	}
+	assertstate.UpdateValidationSet(s.state, &tr)
+
+	si1 := &snap.SideInfo{RealName: "some-snap", SnapID: "some-snap-id", Revision: snap.R(1)}
+	snapstate.Set(s.state, "some-snap", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{si1},
+		Current:  snap.R(1),
+		SnapType: "app",
+	})
+	snaptest.MockSnap(c, `name: some-snap`, si1)
+
+	si2 := &snap.SideInfo{RealName: "some-other-snap", SnapID: "some-other-snap-id", Revision: snap.R(1)}
+	snapstate.Set(s.state, "some-other-snap", &snapstate.SnapState{
+		Active:   true,
+		Sequence: []*snap.SideInfo{si2},
+		Current:  snap.R(1),
+		SnapType: "app",
+	})
+	snaptest.MockSnap(c, `name: some-other-snap`, si2)
+
+	s.fakeBackend.linkSnapFailTrigger = filepath.Join(dirs.SnapMountDir, "/some-other-snap/11")
+
+	names, tss, err := snapstate.UpdateMany(context.Background(), s.state, nil, s.user.ID, &snapstate.Flags{})
+	c.Assert(err, IsNil)
+	c.Check(names, DeepEquals, []string{"some-other-snap", "some-snap"})
+	c.Check(logbuf.String(), Equals, "")
+	chg := s.state.NewChange("update", "")
+	for _, ts := range tss {
+		chg.AddAll(ts)
+	}
+
+	s.settle(c)
+
+	return chg
+}
+
+func (s *validationSetsSuite) TestUpdateManyValidationSetsPartialFailureNothingToRestore(c *C) {
+	var refreshed []string
+	restoreMaybeRestoreValidationSetsAndRevertSnaps := snapstate.MockMaybeRestoreValidationSetsAndRevertSnaps(func(st *state.State, refreshedSnaps []string) ([]*state.TaskSet, error) {
+		refreshed = refreshedSnaps
+		// nothing to restore
+		return nil, nil
+	})
+	defer restoreMaybeRestoreValidationSetsAndRevertSnaps()
+
+	var addCurrentTrackingToValidationSetsStackCalled int
+	restoreAddCurrentTrackingToValidationSetsStack := snapstate.MockAddCurrentTrackingToValidationSetsStack(func(st *state.State) error {
+		addCurrentTrackingToValidationSetsStackCalled++
+		return nil
+	})
+	defer restoreAddCurrentTrackingToValidationSetsStack()
+
+	s.testUpdateManyValidationSetsPartialFailure(c)
+
+	// only some-snap was successfully refreshed, this also confirms that
+	// mockMaybeRestoreValidationSetsAndRevertSnaps was called.
+	c.Check(refreshed, DeepEquals, []string{"some-snap"})
+
+	// validation sets history has been updated
```

Comment:
to be clear, it gets "updated" but because of the revert this is the add is a no-op, right? this comment maybe should be clearer on that

Reply:
Yes, I've expanded the comment, also inside handlers.go.